pkg/controllers/machine/termination/suite_test.go

```diff
@@ -0,0 +1,183 @@
+})
+
+var _ = AfterSuite(func() {
+	Expect(env.Stop()).To(Succeed(), "Failed to stop environment")
+})
+
+var _ = AfterEach(func() {
+	fakeClock.SetTime(time.Now())
+	ExpectCleanedUp(ctx, env.Client)
+	cloudProvider.Reset()
+})
+
+var _ = Describe("Termination", func() {
+	var provisioner *v1alpha5.Provisioner
+	var machine *v1alpha5.Machine
+
+	BeforeEach(func() {
+		provisioner = test.Provisioner()
+		machine = test.Machine(v1alpha5.Machine{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: map[string]string{
+					v1alpha5.ProvisionerNameLabelKey: provisioner.Name,
+				},
+			},
+			Spec: v1alpha5.MachineSpec{
+				Resources: v1alpha5.ResourceRequirements{
+					Requests: v1.ResourceList{
+						v1.ResourceCPU:          resource.MustParse("2"),
+						v1.ResourceMemory:       resource.MustParse("50Mi"),
+						v1.ResourcePods:         resource.MustParse("5"),
+						fake.ResourceGPUVendorA: resource.MustParse("1"),
+					},
+				},
+			},
+		})
+	})
+	It("should delete the node and the CloudProvider Machine when Machine deletion is triggered", func() {
+		ExpectApplied(ctx, env.Client, provisioner, machine)
+		ExpectReconcileSucceeded(ctx, machineController, client.ObjectKeyFromObject(machine))
+
+		machine = ExpectExists(ctx, env.Client, machine)
+		_, err := cloudProvider.Get(ctx, machine.Status.ProviderID)
+		Expect(err).ToNot(HaveOccurred())
+
+		node := test.MachineLinkedNode(machine)
+		ExpectApplied(ctx, env.Client, node)
+
+		// Expect the node and the machine to both be gone
+		Expect(env.Client.Delete(ctx, machine)).To(Succeed())
+		ExpectReconcileSucceeded(ctx, terminationController, client.ObjectKeyFromObject(machine)) // triggers the node deletion
+		ExpectFinalizersRemoved(ctx, env.Client, node)
+		ExpectReconcileSucceeded(ctx, terminationController, client.ObjectKeyFromObject(machine)) // now all nodes are gone so machine deletion continues
+		ExpectNotFound(ctx, env.Client, machine, node)
+
+		// Expect the machine to be gone from the cloudprovider
+		_, err = cloudProvider.Get(ctx, machine.Status.ProviderID)
+		Expect(cloudprovider.IsMachineNotFoundError(err)).To(BeTrue())
+	})
+	It("should delete multiple Nodes if multiple Nodes map to the Machine", func() {
+		ExpectApplied(ctx, env.Client, provisioner, machine)
+		ExpectReconcileSucceeded(ctx, machineController, client.ObjectKeyFromObject(machine))
+
+		machine = ExpectExists(ctx, env.Client, machine)
+		_, err := cloudProvider.Get(ctx, machine.Status.ProviderID)
+		Expect(err).ToNot(HaveOccurred())
+
+		node1 := test.MachineLinkedNode(machine)
+		node2 := test.MachineLinkedNode(machine)
+		node3 := test.MachineLinkedNode(machine)
+		ExpectApplied(ctx, env.Client, node1, node2, node3)
+
+		// Expect the node and the machine to both be gone
+		Expect(env.Client.Delete(ctx, machine)).To(Succeed())
+		ExpectReconcileSucceeded(ctx, terminationController, client.ObjectKeyFromObject(machine)) // triggers the node deletion
```

Comment:
I think we should assert that the nodes are still found before removing finalizers, right?
```suggestion
                ExpectExists(ctx, env.Client, node1, node2, node3)
		ExpectFinalizersRemoved(ctx, env.Client, node)
```

Reply:
Same comment as above